File: ocssw_src/src/nav_seahawk/hawknav/orb2lla.py
def orb2lla(orb):
    # Compute geodetic longitude, latitude and altitude 
    # from Cartesian orbit vector in ECR coordinates
    # Uses geodetic approximation from Patt and Gregg, IJRS, 1993.
    # input: orb: position vector[0,1,2]
    # reference: orb2lla.pro by Fred Patt
    # ported to Python by Liang Hong, 2/13/2020
    
    import numpy as np
    
    RE = 6378.137           # Earth equatorial radius (km)
    REM = 6371.0            # Earth mean radius (km)
    F = 1.0/298.257       # Earth flattening factor
    OMF2 = (1.0-F)**2
        
    # Compute longitude
    lon = np.rad2deg(np.arctan2(orb[:,1],orb[:,0]))
    
    # Compute geodetic latitude 
    rad = np.sqrt(np.square(orb[:,0])+np.square(orb[:,1])+np.square(orb[:,2]))
    omf2p = (OMF2*REM + rad - REM)/rad
    pxy = np.square(orb[:,0])+np.square(orb[:,1])
    temp = np.sqrt(np.square(orb[:,2]) + omf2p*omf2p*pxy)
    lat = np.rad2deg(np.arcsin(np.array(orb[:,2])/temp))

    # Compute altitude
    clatg = np.cos(np.arctan(OMF2*np.tan(np.deg2rad(lat))))
    rl = RE*(1.0-F)/np.sqrt(1.0-(2.0-F)*F*np.square(clatg))
    alt = rad - rl
    
    return lon,lat,alt

File: ocssw_src/src/l0qc_viirs/orb2lla.c
// Compute geodetic longitude, latitude and altitude 
//  from Cartesian orbit vector in ECR coordinates

// Uses geodetic approximation from Patt and Gregg, IJRS, 1993.

//	Arguments
//     
//     	Name    Type 	I/O 	Description
//     	----	---- 	--- 	-----------
//       nRecords   int   		I     number of records
//       orb      	double    	I     3 x nRecords array of orbit vectors
//       lon    	double   	O     nRecords array of longitude
//       lat     	double    	O     nRecords array of latitude
//       alt     	double    	O     nRecords array of altitude

// Liang Hong, Jan 14, 2016, Ported from IDL
// V0.1, Feb 1, 2016

#include <math.h>
#include <genutils.h>

int orb2lla(int nRecords, double orb[], double lon[], double lat[], double alt[]) {
    // Define constants
    double re = 6378.137; // Earth equatorial radius (km)
    double rem = 6371; // Earth mean radius (km)
    double f = 1.0 / 298.257; // Earth flattening factor

    int i;
    double omf2, omf2p, rad, pxy, temp, rl;
    omf2 = (1.0 - f)*(1.0 - f);

    for (i = 0; i < nRecords; i++) {
        // Compute longitude
        lon[i] = (180.0 / OEL_PI) * atan2(orb[1 + i * 6], orb[0 + i * 6]);

        // Compute geodetic latitude 
        rad = sqrt(orb[0 + i * 6] * orb[0 + i * 6] + orb[1 + i * 6] * orb[1 + i * 6] + orb[2 + i * 6] * orb[2 + i * 6]);
        omf2p = (omf2 * rem + rad - rem) / rad;
        pxy = orb[0 + i * 6] * orb[0 + i * 6] + orb[1 + i * 6] * orb[1 + i * 6];
        temp = sqrt(orb[2 + i * 6] * orb[2 + i * 6] + omf2p * omf2p * pxy);
        lat[i] = (180.0 / OEL_PI) * asin(orb[2 + i * 6] / temp);

        // Compute altitude
        rl = re * (1.0 - f) / sqrt(1.0 - (2.0 - f) * f * pxy / (rad * rad));
        alt[i] = rad - rl;
    }
    return 0;
}
